storm-core/src/jvm/backtype/storm/scheduler/resource/strategies/ResourceAwareStrategy.java

```diff
@@ -0,0 +1,478 @@
+        _nodes = RAS_Node.getAllNodesFrom(cluster, _topologies);
+        _availNodes = this.getAvailNodes();
+        this.LOG = LoggerFactory.getLogger(this.getClass());
+        _clusterInfo = cluster.getNetworkTopography();
+        LOG.info(this.getClusterInfo());
+    }
+
+    //the returned TreeMap keeps the Components sorted
+    private TreeMap<Integer, List<ExecutorDetails>> getPriorityToExecutorDetailsListMap(
+            Queue<Component> ordered__Component_list, Collection<ExecutorDetails> unassignedExecutors) {
+        TreeMap<Integer, List<ExecutorDetails>> retMap = new TreeMap<Integer, List<ExecutorDetails>>();
+        Integer rank = 0;
+        for (Component ras_comp : ordered__Component_list) {
+            retMap.put(rank, new ArrayList<ExecutorDetails>());
+            for(ExecutorDetails exec : ras_comp.execs) {
+                if(unassignedExecutors.contains(exec)) {
+                    retMap.get(rank).add(exec);
+                }
+            }
+            rank++;
+        }
+        return retMap;
+    }
+
+    public Map<WorkerSlot, Collection<ExecutorDetails>> schedule(TopologyDetails td) {
+        if (_availNodes.size() <= 0) {
+            LOG.warn("No available nodes to schedule tasks on!");
+            return null;
+        }
+        Collection<ExecutorDetails> unassignedExecutors = _cluster.getUnassignedExecutors(td);
+        Map<WorkerSlot, Collection<ExecutorDetails>> schedulerAssignmentMap = new HashMap<WorkerSlot, Collection<ExecutorDetails>>();
+        LOG.debug("ExecutorsNeedScheduling: {}", unassignedExecutors);
+        Collection<ExecutorDetails> scheduledTasks = new ArrayList<ExecutorDetails>();
+        List<Component> spouts = this.getSpouts(_topologies, td);
+
+        if (spouts.size() == 0) {
+            LOG.error("Cannot find a Spout!");
+            return null;
+        }
+
+        Queue<Component> ordered__Component_list = bfs(_topologies, td, spouts);
+
+        Map<Integer, List<ExecutorDetails>> priorityToExecutorMap = getPriorityToExecutorDetailsListMap(ordered__Component_list, unassignedExecutors);
+        Collection<ExecutorDetails> executorsNotScheduled = new HashSet<ExecutorDetails>(unassignedExecutors);
+        Integer longestPriorityListSize = this.getLongestPriorityListSize(priorityToExecutorMap);
+        //Pick the first executor with priority one, then the 1st exec with priority 2, so on an so forth. 
+        //Once we reach the last priority, we go back to priority 1 and schedule the second task with priority 1.
+        for (int i = 0; i < longestPriorityListSize; i++) {
+            for (Entry<Integer, List<ExecutorDetails>> entry : priorityToExecutorMap.entrySet()) {
+                Iterator<ExecutorDetails> it = entry.getValue().iterator();
+                if (it.hasNext()) {
+                    ExecutorDetails exec = it.next();
+                    LOG.info("\n\nAttempting to schedule: {} of component {}[avail {}] with rank {}",
```

Comment:
Again I think all of these should be debug messages, we don't need them printed out all the time, but we can configure them on if we are doing debugging.